### PR TITLE
Add https functionality

### PR DESCRIPTION
**Pull request by mumme74**
**Add https://**
This implements an option to use https and a way to simply create self signed certificate and private key.
You can of course use a custem key-cert pair if you want.

Many things today need https to be activated, including sensor suite for mobiles.

refs: #113, #8

**My rationale for wanting this pull req, even though chrome apps discontinue:**
I would very much like this to be merged even though apps is not supported after 2022-06
I plan to use this webapp when I'am teaching. Students all have chromebooks and can't find much simpler way to serve https content. Like a webpage that responds to a mobile's accelerometer. 
At least we have a year before its closed down.

Having https in the app in chromestore makes it much easier when students must work on there own at home (due to the pandemic)

### CODE REVIEW

**Comment by kzahel** [comment]
Wow! This is really exciting.
I tried your branch but ran into this problem when trying to load the page.
```
This site can’t provide a secure connection
127.0.0.1 didn’t accept your login certificate, or one may not have been provided.
Try contacting the system admin.
ERR_BAD_SSL_CLIENT_AUTH_CERT
```

**Comment by kzahel** [comment]
Ah, duh. I didn't click on the button to generate the key! A small suggestion, I think it would be an improvement if clicking the "https" checkbox would show/hide the certificate info. That way I would know when clicking it that I need to do something there.

This is awesome! Thanks so much for the PR. I'll test it a little bit more, merge and release soon.

**Comment by kzahel** [comment]
One issue I see (not sure if you see it) is when I click generate crypto, the placeholder label on the key input fields overlays the key, until you type something into the field. Not a blocker, but might be good to see if it's possible to fix somehow

https://user-images.githubusercontent.com/158352/113490063-29c3cd00-94c8-11eb-9417-f81e7828d80d.mp4

File: stream.js [hunk 1]
```diff
@@ -219,13 +227,96 @@
...
+			return serverCert; //WSC.Tls.data.server.cert;
+		  },
+		  getPrivateKey: function(c, cert) {
+			//console.log('Server getting privateKey for \"' + cert + '\"...');
+			return privateKey;//WSC.Tls.data.server.privateKey;
+		  },
+		  tlsDataReady: function(c) {
+			// send TLS data to client
+			var cb = _t.writeCallbacks.pop() || function(){};
+			let str = c.tlsData.getBytes();
```
Comment (kzahel) [inline stream.js:277]: It looks like you fixed a bug with the byte string -> array buffer conversion. I hope this didn't take too much time to debug.

Comment (kzahel) [reply stream.js:277]: Hmm It looks like I had a working version of this for the (unused) websockets stuff
WSC.stringToUint8Array. Anyway, this looks good.

File: makedeps/package.json [hunk 2]
```diff
@@ -7,11 +7,16 @@
     "make": "./node_modules/.bin/browserify index.js > ../assets/bundle.js"
   },
   "dependencies": {
+    "@babel/plugin-transform-react-jsx": "^7.13.12",
```
Comment (kzahel) [inline makedeps/package.json:10]: I'm confused why this is here -- there's no jsx in these deps? all the jsx lives in react-ui?

File: react-ui/js/index.js [hunk 3]
```diff
@@ -84,7 +86,25 @@ const functions = {
...
+    console.log('certificate', val);
+    console.assert(typeof val === 'string')
+    app.webapp.updateOption('optCertificate', val);
+  },
+  optUseHttps: (app, k, val) => {
+    console.log("useHttps", val);
+    app.webapp.updateOption('optUseHttps', val);
+    if (app.webapp.started) {
+      app.webapp.stop();
+      app.webapp.start();
```
Comment (kzahel) [inline react-ui/js/index.js:105]: this is helpful to stop/start 👍

File: stream.js [hunk 4]
```diff
@@ -219,13 +227,96 @@
...
+			var b = WSC.str2ab(str);
+			//console.log('encrypt to client: ' + str);
+			if (this.connected)
+                chrome.sockets.tcp.send( _t.sockId, b, cb);
+            else
+                _t.error("tlsData on closed socket");
+		  },
```
Comment (kzahel) [inline stream.js:284]: diff hard to read. it looks like you use tabs and not spaces :-) I don't really care though

File: webapp.js [hunk 5]
```diff
@@ -1,12 +1,12 @@
 (function(){
-    var sockets = chrome.sockets
 
     function WebApplication(opts) {
         // need to support creating multiple WebApplication...
         if (WSC.DEBUG) {
             console.log('initialize webapp with opts',opts)
         }
         opts = opts || {}
+        //opts.optUseHttps = true; /// FIXME hack this into a gui selectable thing
```
Comment (kzahel) [inline webapp.js:9]: remove

File: makedeps/package.json (hunk 2 above)
Comment (mumme74) [reply makedeps/package.json:10]: Sorry might, be a leftover from my babysteps into this project.
I will remove..

File: stream.js (hunk 1 above)
Comment (mumme74) [reply stream.js:277]: The whole thing was kind of a steep learning curve, fun but hard.
tried WSC.stringToUint8Array but couldn't get it to work. It scrambled output if I recall correctly.

File: stream.js (hunk 4 above)
Comment (mumme74) [reply stream.js:284]: I am working an a patch with this and your your other suggestions.
Hopefully be done soon enough.

**Comment by mumme74** [comment]
### Thank you for the quick response and all the suggested inprovements!

I have now done the suggested changes, and implement hide/show https fields. 
I have never used react ui before so I hope I didn't do something obviously wrong.

Regarding the textbox hint issue in your movie, I can't see that one here. But I think I found a workaround for a possible cause.

Very glad to hear that you are willing to merge this and update appstore! :+1: :grinning:


**I hope that the code now is in a mergable state with my last push!** #3e20c13

**Review by kzahel — APPROVED** [review]
Thanks for the changes! looks great.

**Comment by kzahel** [comment]
Just submitted the new version for review! If you see version 0.5.2 in the store it should have https working 🤞

**Comment by mumme74** [comment]
Perfect!

Thank you!

**Comment by kzahel** [comment]
It's in the store and I just tested it and it works on my chromebook.
Interestingly on a Mac device it doesn't have the "Proceed ... (unsafe)" button. But after looking on stack overflow it looks like there's a "secret code" if you type

"thisisunsafe" into the error page, it allows you to see the content.

Haha. I wonder if it's worthwhile mentioning that somewhere near the "enable https" button (if they're on a mac at least) 🤷

**Comment by mumme74** [comment]
Nice ! :smile:

Regarding MAC, yes perhaps you are right, I would have to try when I get hold of a MAC. 
Is it safari or chrome or both on MAC? Whole thing seems strange :raised_eyebrow:

It might be related to node-forge not supporting TLS 1.2 :interrobang:  :thinking:
I am working on a patch for node-forge to support TLS1.2 although not yet finished. :roll_eyes:
https://github.com/mumme74/forge/tree/tls12

**Pull request merged**